src/cfg/cfg-traversal.h

```diff
@@ -304,16 +341,20 @@ struct CFGWalker : public ControlFlowWalker<SubType, VisitorType> {
         break;
       }
       case Expression::Id::TryId: {
-        // FIXME Update the implementation to match the new spec
-        WASM_UNREACHABLE("unimp");
-        /*
         self->pushTask(SubType::doEndTry, currp);
-        self->pushTask(SubType::scan, &curr->cast<Try>()->catchBody);
-        self->pushTask(SubType::doStartCatch, currp);
+        auto& catchBodies = curr->cast<Try>()->catchBodies;
+        using namespace std::placeholders;
+        for (Index i = 0; i < catchBodies.size(); i++) {
+          auto doEndCatchI = std::bind(doEndCatch, _1, _2, i);
+          self->pushTask(doEndCatchI, currp);
+          self->pushTask(SubType::scan, &catchBodies[i]);
+          auto doStartCatchI = std::bind(doStartCatch, _1, _2, i);
+          self->pushTask(doStartCatchI, currp);
```

Comment:
Asan says this line causes a memory leak, but I'm not very sure why... Do you have a guess?

Reply:
Very odd, no idea... code looks 100% valid to me.

Reply:
I think I figured out what's going on. Surprisingly `std::function`'s constructor does some dynamic allocation in some cases. (It seems this behavior is dropped in C++17, but we are using C++14.) And out `SmallVector` does not call destructors of objects on `pop_back()` of `clear()` when they are stored in `fixed` part. (`flexible` part is fine because it's using `std::vector`.) It was fine until now because what we've stored in `SmallVector`, such as `Expression`s, are all managed by a separate allocation arena.

Fixing this does not seem to be trivial. We can't call destructors on all values (there are primitive types), so `SmallVector` has to be somewhat complicated. And first of all, given that `std::function` can cause a dynamic allocation, I'm not sure if we should use it in the part that can be performance critical. The reason I changed 
```cpp
typedef void (*TaskFunc)(SubType*, Expression**);
```
to
```cpp
using TaskFunc = std::function<void(SubType*, Expression**)>;
```
was the result of `std::bind` can be passed as `std::function` but not a function pointer. The reason I used `std::bind` and not a lambda was a lambda capturing variables cannot be passed as a function argument. (Here we need to capture `i`)

Maybe we should just use lambdas and make `i` a class variable or something, which seems less cleaner, but maybe we can't help..

Reply:
Oh wow, good find... C++ is annoying sometimes.

Yes, lambdas seem best for now, I can't think of a better option.

Reply:
I changed it to lambdas:  https://github.com/WebAssembly/binaryen/pull/3494#discussion_r561012770
I was a little confused. I thought capturing lambdas couldn't be passed as a function argument but that turned out to be only the case when we use the function pointer as `TaskFunc`. After we changed it to `std::function` it works fine. Also, not sure why, but the asan memory leak error does not appear in the CG anymore 🤷🏻 (Was that linked to a condition related to `std::bind`? No idea) So I guess we will go with this.

About possible performance issues, I ran all wasm-opt tests with 1. `TaskFunc` being a function pointer and 2. `TaskFunc` being a `std::function`, in release build. The performance difference doesn't seem to be very concerning. WDYT? All numbers are measured three times but the first result was thrown out of concerns that it might be affected by the cache warming up.

- `TaskFunc` is a function pointer
real    0m56.546s
user    0m46.372s
sys     1m20.068s

real    0m57.272s
user    0m46.733s
sys     1m21.784s

- `TaskFunc` is a `std::function`
real    0m58.454s
user    0m47.954s
sys     1m22.177s

real    0m57.924s
user    0m47.401s
sys     1m22.224s